src/ray/object_manager/object_manager.cc

```diff
@@ -0,0 +1,460 @@
+}
+
+void ObjectManager::HandlePushReceive(TCPClientConnection::pointer conn, BoostEC length_ec){
+  std::vector<uint8_t> message;
+  message.resize(conn->message_length_);
+  boost::system::error_code ec;
+  boost::asio::read(conn->GetSocket(), boost::asio::buffer(message), ec);
+  // Serialize.
+  auto object_header = flatbuffers::GetRoot<ObjectHeader>(message.data());
+  ObjectID object_id = ObjectID::from_binary(object_header->object_id()->str());
+  int64_t object_size = (int64_t) object_header->object_size();
+  int64_t metadata_size = 0;
+  // Try to create shared buffer.
+  std::shared_ptr<Buffer> data;
+  arrow::Status s = store_client_->GetClient().Create(object_id.to_plasma_id(), object_size, NULL, metadata_size, &data);
+  if(s.ok()){
+    // Read object into store.
+    uint8_t *mutable_data = data->mutable_data();
+    boost::asio::read(conn->GetSocket(), boost::asio::buffer(mutable_data, object_size), ec);
+    if(!ec.value()){
+      ARROW_CHECK_OK(store_client_->GetClient().Seal(object_id.to_plasma_id()));
+      ARROW_CHECK_OK(store_client_->GetClient().Release(object_id.to_plasma_id()));
+    } else {
+      ARROW_CHECK_OK(store_client_->GetClient().Release(object_id.to_plasma_id()));
+      ARROW_CHECK_OK(store_client_->GetClient().Abort(object_id.to_plasma_id()));
+      cout << "Receive Failed" << endl;
+    }
+  } else {
+    cout << "Buffer Create Failed: " << s.message() << endl;
+    // Read object into empty buffer.
+    uint8_t *mutable_data = (uint8_t *) malloc(object_size + metadata_size);
+    boost::asio::read(conn->GetSocket(), boost::asio::buffer(mutable_data, object_size), ec);
+  }
+  // Wait for another push.
+  ray::Status ray_status = WaitPushReceive(conn);
+};
+
+ray::Status ObjectManager::QueuePush(const ObjectID &object_id_const,
+                                     SenderConnection::pointer conn){
+  ObjectID object_id = ObjectID(object_id_const);
+  if(conn->ObjectIdQueued(object_id)){
+    // For now, return with status OK if the object is already in the send queue.
+    return ray::Status::OK();
+  }
+  conn->QueueObjectId(object_id);
+  if(num_transfers_ < max_transfers_){
+    return ExecutePushQueue(conn);
+  }
+  return ray::Status::OK();
+};
+
+ray::Status ObjectManager::ExecutePushQueue(SenderConnection::pointer conn){
+  ray::Status status = ray::Status::OK();
+  while(num_transfers_ < max_transfers_){
+    if (conn->IsObjectIdQueueEmpty()){
+      return ray::Status::OK();
+    }
+    ObjectID object_id = conn->DeQueueObjectId();
+    // The threads that increment/decrement num_transfers_ are different.
+    // It's important to increment num_transfers_ before executing the push.
+    num_transfers_ += 1;
+    status = ExecutePushMeta(object_id, conn);
+  }
+  return status;
+};
+
+ray::Status ObjectManager::ExecutePushMeta(const ObjectID &object_id_const,
+                                       SenderConnection::pointer conn) {
+  ObjectID object_id = ObjectID(object_id_const);
```

Comment:
So are we sending metadata and data separately?

Reply:
These are the headers (renamed accordingly).